### PR TITLE
ns for fx: remove restriction on nodes with no args and only kwargs

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#78181 ns for fx: remove restriction on nodes with no args and only kwargs**

Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

### CODE REVIEW

File: torch/ao/ns/fx/graph_passes.py [hunk 1]
```diff
@@ -27,6 +27,33 @@
...
+            assert len(norm_args) + len(norm_kwargs) >= idx + 1
+            if idx < len(norm_args):
+                return norm_args[idx]
+            else:
+                # note: in Python 3.7+ dicts are ordered
+                return list(norm_kwargs.values())[idx]
+        else:
+            assert len(node.args) > idx
+            return node.args[idx]  # type: ignore[return-value]
+    except RuntimeError:
```
Comment (z-a-f) [inline torch/ao/ns/fx/graph_passes.py:49]: nit: Can we add a comment what use case this exception is catching?

**Review by z-a-f — APPROVED** [review]
sgtm

File: torch/ao/ns/fx/graph_passes.py [hunk 2]
```diff
@@ -27,6 +27,33 @@
...
+    """
+    Given a node, gets the n'th input to that node, normalizing
+    args and kwargs to the best of its ability.
+    """
+    try:
+        norm_args_and_kwargs = node.normalized_arguments(
+            gm, normalize_to_only_use_kwargs=True)
+        if norm_args_and_kwargs is not None:
+            norm_args, norm_kwargs = norm_args_and_kwargs
+            assert len(norm_args) + len(norm_kwargs) >= idx + 1
```
Comment (jerryzh168) [inline torch/ao/ns/fx/graph_passes.py:40]: why is this `idx` + 1 and L50 is `idx`? can we have a comment?

Comment (jerryzh168) [reply torch/ao/ns/fx/graph_passes.py:40]: oh this is `>=` and the other one is `>`, maybe better to use the same?

File: torch/ao/ns/fx/graph_passes.py [hunk 3]
```diff
@@ -396,14 +426,40 @@ def _can_insert(node_a_arg, gm_a):
 
     # For each node, check if we handle the copy behavior. This follows the
     # logic in `_insert_copy_of_subgraph_a_after_input_node_c`.
-    for node_a_arg in nodes[0].args[num_non_param_args_node_a:]:
-        if not _can_insert(node_a_arg, gm_a):
-            return False
+    for node_a in nodes:
```
Comment (jerryzh168) [inline torch/ao/ns/fx/graph_passes.py:429]: is it possible to have a function to return `nodes[num:]`?

Comment (vkuzo) [reply torch/ao/ns/fx/graph_passes.py:436]: yeah, I looked into it, didn't see a way that was significantly better than the current code

**Review by hx89 — APPROVED** [review]

**Pull request closed without merging**